### DISCUSSION TITLE
What is the case when ERC20.transferFrom(...) returns false

### DISCUSSION THREAD

**Discussion by asadhayat1068** [Cyfrin/foundry-full-course-cu#2664] [Q&A] [answered]
So I am writing test from DSCEngine in Stable Coin Lesson.
I am writing unit test for the `depositCollateral(..)` function.

I want to write a test for the case when this function will revert with `DSCEngine__DepositCollateralFailed` error.
This will only happen when `bool success = IERC20(_tokenCollateralAddress).transferFrom(msg.sender, address(this), _amountCollateral);` returns false.

***How can I make the `ERC20.transferFrom(..)` return false.***

```  function depositCollateral(address _tokenCollateralAddress, uint256 _amountCollateral)
        public
        moreThanZero(_amountCollateral)
        isAllowedToken(_tokenCollateralAddress)
        nonReentrant
    {
        s_collateralDeposited[msg.sender][_tokenCollateralAddress] += _amountCollateral;
        emit CollateralDeposited(msg.sender, _tokenCollateralAddress, _amountCollateral);

        bool success = IERC20(_tokenCollateralAddress).transferFrom(msg.sender, address(this), _amountCollateral);
        if (!success) {
            revert DSCEngine__DepositCollateralFailed(); // <---- **_This Revert_**
        }
    }```

**Comment by EngrPips** [comment, marked as answer]
Hello, You need to create an ERC20 contract that returns false on `transfer, and make the ERC20 token you created one of the collateral tokens of the DSEngine. That means the test you want to write requires a setUp of it own. Does all of this makes snse to you?

**Reply by asadhayat1068** [reply]
I had this approach in mind but thought may be there is a case in which `ERC20.transferFrom()` might return false.

**Reply by EngrPips** [reply]
I don't think they naturally return a boolean `false` but i think they just revert instead but maybe I am wrong.
